File: dilation-code/scripts/TimeKeeper_functions.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/syscall.h>
#include <stdarg.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <linux/netlink.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/select.h>
#include "TimeKeeper_functions.h"
#include "TimeKeeper_definitions.h"
#include "utility_functions.h"


#define SIG_END 44

int select_dialated(int nfds, fd_set *read_fds, fd_set *write_fds, fd_set *except_fds, struct timeval *timeout){

#ifdef __x86_64
	return syscall(316,nfds,read_fds,write_fds,except_fds,timeout);
	
#endif
	return syscall(353,nfds,read_fds,write_fds,except_fds,timeout);
	

}

/*
The original, unmodified gettimeofday() system call
*/
void gettimeofdayoriginal(struct timeval *tv, struct timezone *tz) {
#ifdef __x86_64
	syscall(314, tv, tz);
	return;
#endif
	syscall(351, tv, tz);
	return;
}

/*
Takes an integer (the time_dilation factor) and returns the time dilated process.
    ie: time_dilation = 2 means that when 4 seconds happens in real life,
		the process only thinks 2 seconds has happened
	time_dilation = 1/2 means that when 2 seconds happens in real life,
		 the process thinks 4 seconds actually happened
*/
int clone_time(unsigned long flags, double dilation, int should_freeze) {
	int pid;
        pid = syscall(SYS_clone, flags, NULL, NULL, NULL, NULL);
	dilate_all(pid, dilation);
	if (should_freeze == 1) {
		freeze_all(pid);
	}
	return 0;
}

/*
Returns the virtual time of an LXC, given it's pid
*/
int gettimepid(int pid, struct timeval *tv, struct timezone *tz) {
#ifdef __x86_64
	return syscall(315, pid, tv, tz);
#endif
	return syscall(352, pid, tv , tz);
}

/*
Returns the virtual time of an LXC, given it's name
*/
int gettimename(char *lxcname, struct timeval *tv, struct timezone *tz) {
	int pid;
	if (is_root()) {
	    pid = getpidfromname(lxcname);
            if (pid == -1) {
	         printf("LXC with name: %s not found\n",lxcname);
                 return -1;
	    }
	    return gettimepid(pid, tv, tz);
	}
	return -1;
}

//Given a frozen process specified by PID, will advance it's virtual time by interval (microseconds)
int leap(int pid, int interval) {
	if (is_root() && isModuleLoaded()) {
		char command[100];
		if (interval > 0) {
			sprintf(command, "%c,%d,%d", LEAP, pid, interval);
			if (send_to_timekeeper(command) == -1)
				return -1;
			return 0;
		}
	}
	return -1;
}

/*
Given a pid, add that container to an experiment. If the timeline is less than 0, add to a CBE experiment
else, it represents a specific timeline
*/
int addToExp(int pid, int timeline) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
		if (timeline < 0) {
        	        sprintf(command, "%c,%d", ADD_TO_EXP_CBE, pid);
		}
		else {
			sprintf(command, "%c,%d,%d", ADD_TO_EXP_CS, pid, timeline);
		}
		if (send_to_timekeeper(command) == -1)
			return -1;
        	return 0;
	}
        return -1;
}

/*
Starts a CBE Experiment
*/
int startExp() {
	if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c", START_EXP);
		send_to_timekeeper(command);
                return 0;
        }
        return 1;
}

/*
Given all Pids added to experiment, will set all their virtual times to be the same, then freeze them all (CBE and CS)
*/
int synchronizeAndFreeze() {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c", SYNC_AND_FREEZE);
                if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

/*
Set the interval in which a pid in a given timeline should advance (microsends) (CS)
*/
int setInterval(int pid, int interval, int timeline) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c,%d,%d,%d", SET_INTERVAL, pid, interval, timeline);
		if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}


/*
Progress all containers in the given timeline to advance by their specified intervals. The function will return when all containers
have done so. (CS)
force = 0, do not force LXC times
force = 1, force LXC times to be progress exactly 
*/
struct msghdr msg;
int progress(int timeline, int force) {
    struct sockaddr_nl src_addr;
    int sock_fd;
    fd_set readfds;
    struct timeval tv;
    int n, rv;

    sock_fd = socket(PF_NETLINK, SOCK_RAW, NETLINK_USER);
    if (sock_fd < 0)
        return -1;

    memset(&src_addr, 0, sizeof(src_addr));
    src_addr.nl_family = AF_NETLINK;
    src_addr.nl_pid = syscall(SYS_gettid);

    bind(sock_fd, (struct sockaddr *)&src_addr, sizeof(src_addr));


    char command[100];
    sprintf(command, "%c,%d,%d,%d", PROGRESS, timeline, gettid(), force);
    if (send_to_timekeeper(command) == -1)
	return -1;

    FD_ZERO(&readfds);
    FD_SET(sock_fd, &readfds);

    n = sock_fd + 1;
    tv.tv_sec = 5; //set for 5 seconds currently, should probably change this
    tv.tv_usec = 0;
// I dont think a timeout should be necessary anymore.
    rv = select(n, &readfds, NULL, NULL, &tv);

    if (rv == -1) {
	perror("select error");
    }
  else if (rv == 0) {
	printf("Timeout occurred, fix timeline..\n");
	sprintf(command, "%c,%d", FIX_TIMELINE, timeline);
	send_to_timekeeper(command);
	sleep(1);
    }
    else {
        recvmsg(sock_fd, &msg, 0);
    }
    close(sock_fd);
	return 0;

}

/*
Reset all pre-specifed intervals for a given timeline (CS)
*/
int reset(int timeline) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c,%d", RESET, timeline);
                if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

/*
Stop a running experiment (CBE or CS) **Do not call stopExp if you are waiting for a s3fProgress to return!!**
*/
int stopExp() {
	if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c", STOP_EXP);
		if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

/*
Sets the TDF of the given pid
*/
int dilate(int pid, double dilation) {
	if (is_root() && isModuleLoaded()) {
		char command[100];
		int dil;
		if ( (dil = fixDilation(dilation)) == -1) {
			return -1;
		}
		printf("Trying to create dilation %d from %f\n",dil, dilation);
		if (dil < 0) {
		sprintf(command, "%c,%d,1,%d", DILATE, pid, dil*-1);
		}
		else {
		sprintf(command, "%c,%d,%d", DILATE, pid, dil);
		}
		if (send_to_timekeeper(command) == -1)
			return -1;
		return 0;
	}
	return -1;
}


/*
Will set the TDF of a LXC and all of its children
*/
int dilate_all(int pid, double dilation) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
		int dil;
		if ( (dil = fixDilation(dilation)) == -1) {
			return -1;
		}
		printf("Trying to create dilation %d from %f\n",dil, dilation);
                if (dil < 0) {
                	sprintf(command, "%c,%d,1,%d", DILATE_ALL, pid, dil*-1);
                }
                else {
                	sprintf(command, "%c,%d,%d", DILATE_ALL, pid, dil);
                }
		if (send_to_timekeeper(command) == -1)
			return -1;
		return 0;
        }
        return -1;
}

/*
Takes an integer (pid of the process). This function will essentially 'freeze' the
time of the process. It does this by sending a sigstop signal to the process.
*/
int freeze(int pid) {
	if (is_root() && isModuleLoaded()) {
		char command[100];
		sprintf(command, "%c,%d,%d", FREEZE_OR_UNFREEZE, pid, SIGSTOP);
		send_to_timekeeper(command);
		return 1;
	}
	return -1;
}

/*
Takes an integer (pid of the process). This function will unfreeze the process.
When the process is unfrozen, it will think that no time has passed, and will
continue doing whatever it was doing before it was frozen.
*/
int unfreeze(int pid) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c,%d,%d", FREEZE_OR_UNFREEZE, pid, SIGCONT);
		if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

/*
Same as freeze, except that it will freeze the process as well as all of its children.
*/
int freeze_all(int pid) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c,%d,%d", FREEZE_OR_UNFREEZE_ALL, pid, SIGSTOP);
		if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

/*
Same as unfreeze, except that it will unfreeze the process as well as all of its children.
*/
int unfreeze_all(int pid) {
        if (is_root() && isModuleLoaded()) {
                char command[100];
                sprintf(command, "%c,%d,%d", FREEZE_OR_UNFREEZE_ALL, pid, SIGCONT);
		if (send_to_timekeeper(command) == -1)
			return -1;
                return 0;
        }
        return -1;
}

File: dilation-code/tests/timekeeper_functions.py
#
# File  : timekeeper_functions.py
# 
# Brief : Exposes TimeKeeper API in python
#
# authors : Vignesh Babu
#


import os
from signal import SIGSTOP, SIGCONT
import time
import subprocess

TIMEKEEPER_FILE_NAME = "/proc/dilation/status"
DILATE = 'A'
DILATE_ALL = 'B'
FREEZE_OR_UNFREEZE = 'C'
FREEZE_OR_UNFREEZE_ALL = 'D'
LEAP = 'E'

ADD_TO_EXP_CBE = 'F'
ADD_TO_EXP_CS = 'G'
SYNC_AND_FREEZE = 'H'
START_EXP = 'I'
SET_INTERVAL = 'J'
PROGRESS = 'K'
FIX_TIMELINE = 'L'
RESET = 'M'
STOP_EXP = 'N'
SET_CBE_EXP_TIMESLICE = 'T'
SET_NETDEVICE_OWNER  = 'U'
PROGRESS_EXP_CBE = 'V'
RESUME_CBE = 'W'






def is_root() :

	if os.geteuid() == 0 :
		return 1
	print "Needs to be run as root"
	return 0


def is_Module_Loaded() :

	if os.path.isfile(TIMEKEEPER_FILE_NAME) == True:
		return 1
	else :
		print "Timekeeper module is not loaded"
		return 0


# Associates the given pid with a net_device
def set_netdevice_owner(pid,dev_name) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR in Set Netdevice Owner. TimeKeeper is not loaded!"
		return -1
	cmd = SET_NETDEVICE_OWNER + "," + str(pid) + "," + str(dev_name)

	return send_to_timekeeper(cmd)




# Converts the double into a integer, makes computation easier in kernel land
# ie .5 is converted to -2000, 2 to 2000

def fixDilation(dilation) :
        
	dil = 0

	if dilation < 0 :
		print "Negative dilation does not make sense";
		return -1
        
	if dilation < 1.0 and dilation > 0.0 :
		dil = (int)((1/dilation)*1000.0)
		dil = dil*-1
        
	elif dilation == 1.0 or dilation == -1.0 :
                dil = 0
	else:
		dil = (int)(dilation*1000.0)
        
        return dil;


def getpidfromname(lxcname) :

	my_pid = os.getpid()
	pid = -1
	temp_file = "/tmp/" + str(my_pid) + "_timekeeper.txt"

	# For different versions of LXCs
	command = "lxc-info -n %s | grep PID | tr -s ' ' | cut -d ' ' -f 2 > %s" % (lxcname,temp_file)
	os.system(command)
	

	with open(temp_file,"r") as f :
		pidstr = f.read()
		if pidstr != "" :
			pid = int(pidstr)
	if pid == -1 :
		command  = "lxc-info -n %s | grep pid | tr -s ' ' | cut -d ' ' -f 2 > %s" % (lxcname,temp_file)
        	os.system(command)
		with open(temp_file,"r") as f :
	                pidstr = f.read()
        	        if pidstr != "" :
                	        pid = int(pidstr)

	
	
	os.remove(temp_file)
	
	return pid

def send_to_timekeeper(cmd) :

	if is_root()== 0 or is_Module_Loaded() == 0 :
		print "ERROR sending cmd to timekeeper"
		return -1

	with open(TIMEKEEPER_FILE_NAME,"w") as f :
		f.write(cmd)
	return 1

# timeslice in nanosecs
def set_cbe_experiment_timeslice(timeslice) :


	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR setting timeslice value"
		return -1
	
	timeslice = int(timeslice)
	cmd = SET_CBE_EXP_TIMESLICE + "," + str(timeslice)

	return send_to_timekeeper(cmd)


#
#Given a pid, add that container to a CBE experiment.
#

def addToExp(pid) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		return -1 
	cmd = ADD_TO_EXP_CBE + "," + str(pid)
	return send_to_timekeeper(cmd)



#
#Starts a CBE Experiment
#

def startExp():

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR starting CBE Experiment"
		return -1
	cmd = START_EXP
	return send_to_timekeeper(cmd)


#
#Given all Pids added to experiment, will set all their virtual times to be the same, then freeze them all (CBE and CS)
#

def synchronizeAndFreeze() :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR in Synchronize and Freeze"
		return -1
	cmd = SYNC_AND_FREEZE
	return send_to_timekeeper(cmd)





#
#Stop a running experiment (CBE or CS) **Do not call stopExp if you are waiting for a s3fProgress to return!!**
#

def stopExp() :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR in Synchronize and Freeze"
		return -1
	cmd = STOP_EXP
	return send_to_timekeeper(cmd)



#
#Sets the TDF of the given pid
#

def dilate(pid,dilation) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR in setting dilation for pid :", pid
		return -1
	dil = fixDilation(dilation)
	if dil == - 1 :
		return - 1
	elif dil < 0 :
		cmd = DILATE + "," + str(pid) + ",1," + str(-1*dil)
	else :
		cmd = DILATE + "," + str(pid) + "," + str(dil)

	return send_to_timekeeper(cmd)

#
# Will set the TDF of a LXC and all of its children
#

def dilate_all(pid,dilation) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		print "ERROR in setting dilation for pid :", pid
		return -1
	dil = fixDilation(dilation)
	if dil == - 1 :
		return - 1
	elif dil < 0 :
		cmd = DILATE_ALL + "," + str(pid) + ",1," + str(-1*dil)
	else :
		cmd = DILATE_ALL + "," + str(pid) + "," + str(dil)

	return send_to_timekeeper(cmd)


#
# Takes an integer (pid of the process). This function will essentially 'freeze' the
# time of the process. It does this by sending a sigstop signal to the process.
#

def freeze(pid) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		return -1
	else :
		cmd = FREEZE_OR_UNFREEZE + "," + str(pid) + "," + SIGSTOP
		return send_to_timekeeper(cmd)

	

#
# Takes an integer (pid of the process). This function will unfreeze the process.
# When the process is unfrozen, it will think that no time has passed, and will
# continue doing whatever it was doing before it was frozen.
#

def unfreeze(pid) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		return -1
	cmd = FREEZE_OR_UNFREEZE + "," + str(pid) + "," +SIGCONT
	return send_to_timekeeper(cmd)




#
# Same as freeze, except that it will freeze the process as well as all of its children.
#

def freeze_all(pid) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		return -1
	else :
		cmd = FREEZE_OR_UNFREEZE_ALL + "," + str(pid) + "," + SIGSTOP
		return send_to_timekeeper(cmd)

	


# 
# Same as unfreeze, except that it will unfreeze the process as well as all of its children.
#

def unfreeze(pid) :

	if is_root() == 0 or is_Module_Loaded() == 0 :
		return -1
	cmd = FREEZE_OR_UNFREEZE_ALL + "," + str(pid) + "," +SIGCONT
	return send_to_timekeeper(cmd)


		
## CS Experiment Fuctions


def add_To_CS_Exp(pid,timeline) :
	if is_root() and is_Module_Loaded() and timeline >= 0 :
		cmd = ADD_TO_EXP_CS + "," + str(pid) + "," + str(timeline)
		return send_to_timekeeper(cmd)
	else:
		return -1


def leap(pid,interval) :
	if is_root() and is_Module_Loaded() and interval > 0 :
		cmd = LEAP + "," + str(pid) + "," + str(interval)
		return send_to_timekeeper(cmd)
	else:
		return -1


def set_interval(pid, interval, timeline) :
	if is_root() and is_Module_Loaded() and interval > 0:
		cmd = SET_INTERVAL + "," + str(pid) + "," + str(interval) + "," + str(timeline)
		return send_to_timekeeper(cmd)
	else:
		return -1


def fix_timeline(timeline) :
	if is_root() and is_Module_Loaded() and timeline >= 0:
		cmd = FIX_TIMELINE + "," + str(timeline)
		return send_to_timekeeper(cmd)
	else:
		return -1


def reset_timeline(timeline) :
	if is_root() and is_Module_Loaded() and timeline >= 0:
		cmd = RESET + "," + str(timeline)
		return send_to_timekeeper(cmd)
	else:
		return -1



def progress(timeline,mypid, force) :
	if is_root() and is_Module_Loaded() and timeline >= 0:
		cmd = PROGRESS + "," + str(timeline) + "," + str(mypid) + "," + str(force)
		return send_to_timekeeper(cmd)
	else:
		return -1

def progress_exp_cbe(n_rounds) :
	if is_root() and is_Module_Loaded() and n_rounds > 0 :
		cmd = PROGRESS_EXP_CBE + "," + str(n_rounds)
		return send_to_timekeeper(cmd)
	else:
		return -1

def resume_exp_cbe() :
	if is_root() and is_Module_Loaded() :
		cmd = RESUME_CBE
		return send_to_timekeeper(cmd)
	else:
		return -1


